modules/optimaticBidAdapter.js

```diff
@@ -0,0 +1,128 @@
+
+export const spec = {
+  code: 'optimatic',
+
+  supportedMediaTypes: ['video'],
+
+  isBidRequestValid: function(bid) {
+    let isValid = !!(bid && bid.params && bid.params.placement && bid.params.bidfloor);
+    if (!isValid) {
+      let error = (!bid) ? 'nobid' : (!bid.params) ? 'nobidparams' : (!bid.params.placement) ? 'nobidplacement' : (!bid.params.bidfloor) ? 'nobidfloor' : 'noerror';
+      let id = (bid && bid.params && bid.params.placement);
+      log('bid-error', '' + id, '', '' + error, 10);
+    }
+    return isValid;
+  },
+
+  buildRequests: function(bids) {
+    return bids.map(bid => {
+      return {
+        method: 'POST',
+        url: ENDPOINT + bid.params.placement,
+        data: getData(bid),
+        options: {contentType: 'application/json'},
+        bidRequest: bid
+      }
+    })
+  },
+
+  interpretResponse: function(response, { bidRequest }) {
+    let bid;
+    let size;
+    let bidResponse;
+    try {
+      response = response.body;
+      bid = response.seatbid[0].bid[0];
+    } catch (e) {
+      response = null;
+    }
+    if (!response || !bid || !bid.adm || !bid.price) {
+      utils.logWarn(`No valid bids from ${spec.code} bidder`);
+      let error = (!response) ? 'nullresponse' : (!bid) ? 'nullbid' : (!bid.adm) ? 'nulladm' : (!bid.price) ? 'nullprice' : 'nullerror';
+      log('bid-error', '' + bidRequest.params.placement, '', '' + error, 10);
+      return [];
+    }
+    size = getSize(bidRequest.sizes);
+    bidResponse = {
+      requestId: bidRequest.bidId,
+      bidderCode: spec.code,
+      cpm: bid.price,
+      creativeId: bid.id,
+      vastXml: bid.adm,
+      width: size.width,
+      height: size.height,
+      mediaType: 'video',
+      currency: response.cur,
+      ttl: 300,
+      netRevenue: true
+    };
+    return bidResponse;
+  }
+};
+
+function getSize(sizes) {
+  let parsedSizes = utils.parseSizesInput(sizes);
+  let [ width, height ] = parsedSizes.length ? parsedSizes[0].split('x') : [];
+  return {
+    width: parseInt(width, 10) || undefined,
+    height: parseInt(height, 10) || undefined
+  };
+}
+
+function getData (bid) {
+  let size = getSize(bid.sizes);
+  let loc = utils.getTopWindowLocation();
+  let global = (window.top) ? window.top : window;
+  return {
+    id: utils.generateUUID(),
+    imp: [{
+      id: '1',
+      bidfloor: bid.params.bidfloor,
+      video: {
+        mimes: ['video/mp4', 'video/ogg', 'video/webm', 'video/x-flv', 'application/javascript', 'application/x-shockwave-flash'],
+        width: size.width,
+        height: size.height
+      }
+    }],
+    site: {
+      id: '1',
+      domain: loc.host,
+      page: loc.href,
+      ref: utils.getTopWindowReferrer(),
+      publisher: {
+        id: '1'
+      }
+    },
+    device: {
+      ua: global.navigator.userAgent,
+      ip: '127.0.0.1',
+      devicetype: 1
+    }
+  };
+}
+
+function log (field1, field2, field3, field4, pct) {
```

Comment:
I see what you're trying to do here... but right now Prebid doesn't have a good way to support things like this. Ajax calls are expensive, and will interfere with both the auction and the publisher's page load performance.

Ideally, we'd offer Adapters some way to "defer" work until the critical page loading stuff is done. For now, could you remove this from the code and add your use-case(s) to #1762?

We're trying to collect use cases for now so we can support it in a more general way.

Reply:
We have made and committed the changes requested.  Please let us know if there is anything else.  Thank you!